File: cpp/lib/ashl/tokenizer.cpp
#include "ashl/tokenizer.hpp"

#include <fstream>
#include <ranges>
#include <sstream>
#include "ashl/utils.hpp"

namespace ashl
{
    std::optional<Token> joinTokensTill(TokenList& tokens, const std::set<std::string>& search)
    {
        if (tokens.Empty()) return {};

        if (search.contains(tokens.Front().value)) return {};

        auto pending = tokens.Front();
        tokens.RemoveFront();
        while (tokens.NotEmpty() && !search.contains(tokens.Front().value))
        {
            auto front = tokens.Front();
            pending = Token{pending.value + front.value, pending.debugInfo + front.debugInfo};
            tokens.RemoveFront();
        }

        return pending;
    }

    bool isSplitToken(const Token& token)
    {
        switch (token.type)
        {
        case TokenType::OpenBrace:
        case TokenType::OpenParen:
        case TokenType::OpenBracket:
        case TokenType::CloseBrace:
        case TokenType::CloseParen:
        case TokenType::CloseBracket:
        case TokenType::Comma:
        case TokenType::StatementEnd:
        case TokenType::Access:
            return true;
        default:
            return false;
        }
    }

    bool isSeparatorToken(const Token& token)
    {
        switch (token.type)
        {
        case TokenType::OpSubtract:
        case TokenType::OpAdd:
        case TokenType::OpDivide:
        case TokenType::OpMultiply:
            return true;
        default:
            return isSplitToken(token) || token.value == " " || token.value == "\n" || token.value == "\r";
        }
    }

    TokenList preprocess(const std::string& fileName, const std::string& fileData)
    {
        TokenList rawTokens{};

        {
            std::stringstream dStream{fileData};
            uint32_t lineNo = 1;
            std::string line{};
            while (std::getline(dStream, line))
            {
                uint32_t colNo = 1;
                for (const char& a : line)
                {
                    rawTokens.InsertBack({std::string{a}, TokenDebugInfo{fileName, lineNo, colNo}});
                    colNo++;
                }
                lineNo++;
            }
        }

        TokenList result{};

        while (rawTokens.NotEmpty())
        {
            auto curToken = rawTokens.Front();

            if (curToken.value == " " || curToken.value == "\n" || curToken.value == "\r")
            {
                rawTokens.RemoveFront();
                continue;
            }

            if (curToken.value == "\"" || curToken.value == "\'")
            {
                auto tok = rawTokens.RemoveFront();
                auto consumed = joinTokensTill(rawTokens, std::set{tok.value});
                rawTokens.RemoveFront();
                if (consumed.has_value())
                {
                    auto consumedTok = consumed.value();

                    result.InsertBack({TokenType::StringLiteral, consumedTok.value, consumedTok.debugInfo});
                }
                continue;
            }

            // if(isSplitToken(curToken))
            // {
            //     rawTokens.RemoveFront();
            //     result.InsertBack(curToken);
            //     continue;
            // }

            if (curToken.value == "/")
            {
                rawTokens.RemoveFront();

                if (rawTokens.NotEmpty())
                {
                    auto nextToken = rawTokens.Front();
                    if (nextToken.value == "/")
                    {
                        rawTokens.RemoveFront();
                        auto startLine = nextToken.debugInfo.startLine;
                        auto combined = rawTokens.RemoveFront();
                        while (rawTokens.NotEmpty() && rawTokens.Front().debugInfo.startLine == startLine)
                        {
                            auto next = rawTokens.RemoveFront();
                            combined = Token{combined.value + next.value, combined.debugInfo + next.debugInfo};
                        }
                        continue;
                    }

                    if (nextToken.value == "*")
                    {
                        rawTokens.RemoveFront();
                        joinTokensTill(rawTokens, std::set<std::string>{"*/"});
                        continue;
                    }
                }

                rawTokens.InsertFront(curToken);
            }

            auto maxSize = std::ranges::reverse_view(Token::SIZES_TO_KEYWORDS).begin()->first;

            std::string combinedStr{};
            std::list<Token> searchTokens{};

            while (combinedStr.size() < static_cast<std::string::size_type>(maxSize) && rawTokens.NotEmpty())
            {
                auto front = rawTokens.Front();
                searchTokens.push_back(rawTokens.RemoveFront());
                combinedStr += front.value;
            }

            auto matchedSize = false;

            for (const auto& [size,matches] : std::ranges::reverse_view(Token::SIZES_TO_KEYWORDS))
            {
                while (static_cast<int>(combinedStr.size()) > size)
                {
                    rawTokens.InsertFront(searchTokens.back());
                    combinedStr = combinedStr.substr(0, combinedStr.size() - searchTokens.back().value.size());
                    searchTokens.pop_back();
                }

                if (static_cast<int>(combinedStr.size()) < size) continue;

                if (matches.contains(combinedStr) && (isSeparatorToken(Token{combinedStr, {}})
                                                          ? true
                                                          : (rawTokens.Empty()
                                                                 ? true
                                                                 : isSeparatorToken(rawTokens.Front()))))
                {
                    matchedSize = true;
                    break;
                }
            }

            if (matchedSize)
            {
                auto debugSpan = searchTokens.front().debugInfo;
                searchTokens.pop_front();
                while (!searchTokens.empty())
                {
                    debugSpan += searchTokens.front().debugInfo;
                    searchTokens.pop_front();
                }

                result.InsertBack({combinedStr, debugSpan});
            }
            else
            {
                if (!searchTokens.empty())
                {
                    while (!searchTokens.empty())
                    {
                        rawTokens.InsertFront(searchTokens.back());
                        searchTokens.pop_back();
                    }
                }

                combinedStr.clear();

                searchTokens.push_back(rawTokens.RemoveFront());
                combinedStr += searchTokens.back().value;
                auto debugSpan = searchTokens.front().debugInfo;

                if (isInteger(searchTokens.front().value))
                {
                    while (rawTokens.NotEmpty() && isInteger(rawTokens.Front().value))
                    {
                        auto tok = rawTokens.RemoveFront();
                        combinedStr += tok.value;
                        debugSpan += tok.debugInfo;
                        searchTokens.push_back(tok);
                    }

                    if (rawTokens.NotEmpty() && rawTokens.Front().value == ".")
                    {
                        auto tok = rawTokens.RemoveFront();
                        combinedStr += tok.value;
                        debugSpan += tok.debugInfo;
                        searchTokens.emplace_back(tok);

                        while (rawTokens.NotEmpty() && isInteger(rawTokens.Front().value))
                        {
                            tok = rawTokens.RemoveFront();
                            combinedStr += tok.value;
                            debugSpan += tok.debugInfo;
                            searchTokens.push_back(tok);
                        }
                    }

                    result.InsertBack(Token{TokenType::Numeric, combinedStr, debugSpan});
                    continue;
                }

                while (rawTokens.NotEmpty() && !isSeparatorToken(rawTokens.Front()))
                {
                    auto tok = rawTokens.RemoveFront();
                    combinedStr += tok.value;
                    debugSpan += tok.debugInfo;

                    searchTokens.push_back(tok);
                }

                result.InsertBack(Token{combinedStr, debugSpan});
                searchTokens.pop_front();
            }
        }

        return result;
    }

    TokenList tokenize(const std::string& fileName, const std::string& fileData)
    {
        return preprocess(fileName, fileData);
    }

    TokenList tokenize(const std::string& fileName)
    {
        std::ifstream fileStream(fileName, std::ios::binary);
        const std::string fileContent((std::istreambuf_iterator<char>(fileStream)),
                                      std::istreambuf_iterator<char>());
        return tokenize(fileName, fileContent);
    }
}

File: csharp/rsl/Tokenizer.cs


namespace rsl;

public class Tokenizer
{
   static Token? JoinTokensTill(ref TokenList tokens, params string[] search)
   {
      if (search.Length == 0) return null;
      if (search.Contains(tokens.Front().Value)) return null;

      var pending = tokens.RemoveFront();
      while (tokens.NotEmpty() && !search.Contains(tokens.Front().Value))
      {
         var front = tokens.RemoveFront();
         pending = new Token(pending.Value + front.Value, pending.DebugInfo + front.DebugInfo);
      }

      return pending;
   }

   static bool IsSplitToken(Token token)
      => token.Type switch
   {
      TokenType.OpenBrace or TokenType.OpenParen or TokenType.OpenBracket or TokenType.CloseBrace
         or TokenType.CloseParen or TokenType.CloseBracket or TokenType.Comma or TokenType.StatementEnd
         or TokenType.Access => true,
      _ => false
   };

   static bool IsSeparatorToken(Token token) => token.Type switch
   {
      TokenType.OpSubtract or TokenType.OpAdd or TokenType.OpDivide or TokenType.OpMultiply or TokenType.Assign  => true,
      _ => IsSplitToken(token) || token.Value is " " or "\n" or "\r"
   };
   
   
   public static TokenList Run(string fileName, string fileData)
   {
      var rawTokens = new TokenList();

      {
         uint lineNo = 1;
         foreach (var line in fileData.SplitLines())
         {
            uint colNo = 1;
            foreach (var col in line)
            {
               rawTokens.InsertBack(new Token(col.ToString(), new DebugInfo(fileName, lineNo, colNo)));
               colNo++;
            }

            lineNo++;
         }
      }


      var result = new TokenList();


      while (rawTokens.NotEmpty())
      {
         var curToken = rawTokens.Front();
         if (curToken.Value is " " or "\n" or "\r")
         {
            rawTokens.RemoveFront();
            continue;
         }
         
         if (curToken.Value is "\"" or "\'")
         {
            var token = rawTokens.RemoveFront();
            var consumed = JoinTokensTill(ref rawTokens, token.Value);
            rawTokens.RemoveFront();
            if (consumed != null)
            {
               result.InsertBack(new Token(TokenType.String, consumed.Value, consumed.DebugInfo));
            }
            continue;
         }

         if (curToken.Value == "/")
         {
            rawTokens.RemoveFront();

            if (rawTokens.NotEmpty())
            {
               var nextToken = rawTokens.Front();
               if (nextToken.Value == "/")
               {
                  rawTokens.RemoveFront();
                  var beginLine = nextToken.DebugInfo.BeginLine;
                  var combined = rawTokens.RemoveFront();
                  while (rawTokens.NotEmpty() && rawTokens.Front().DebugInfo.BeginLine == beginLine)
                  {
                     var next = rawTokens.RemoveFront();
                     combined = new Token(combined.Value + next.Value, combined.DebugInfo + next.DebugInfo);
                  }
                  continue;
               }

               if (nextToken.Value == "*")
               {
                  rawTokens.RemoveFront();
                  JoinTokensTill(ref rawTokens, "*/");
                  continue;
               }
            }

            rawTokens.InsertFront(curToken);
         }

         var maxSize = Token.TokenSizesToKeywords.Keys.Max();

         string combinedString = "";
         var searchTokens = new TokenList();

         while (combinedString.Length < maxSize && rawTokens.NotEmpty())
         {
            var front = rawTokens.Front();
            searchTokens.InsertBack(rawTokens.RemoveFront());
            combinedString += front.Value;
         }

         var matchedSize = false;

         foreach (var (size,matches) in Token.TokenSizesToKeywords.Reverse())
         {
            while (combinedString.Length > size)
            {
               var back = searchTokens.RemoveBack();
               rawTokens.InsertFront(back);
               combinedString = combinedString[..^back.Value.Length];
            }
            
            if(combinedString.Length < size) continue;

            if (matches.Contains(combinedString) && (IsSeparatorToken(new Token(combinedString, new DebugInfo())) ||
                                                     (rawTokens.Empty() || IsSeparatorToken(rawTokens.Front()))))
            {
               matchedSize = true;
               break;
            }
         }

         if (matchedSize)
         {
            var debugSpan = searchTokens.Front().DebugInfo;
            searchTokens.RemoveFront();
            while (searchTokens.NotEmpty())
            {
               debugSpan += searchTokens.Front().DebugInfo;
               searchTokens.RemoveFront();
            }

            result.InsertBack(new Token(combinedString, debugSpan));
         }
         else
         {
            if (searchTokens.NotEmpty())
            {
               rawTokens.InsertFront(searchTokens);
            }

            combinedString = "";

            searchTokens.InsertBack(rawTokens.RemoveFront());
            combinedString += searchTokens.Back().Value;
            var debugSpan = searchTokens.Front().DebugInfo;

            if (int.TryParse(searchTokens.Front().Value, out var _))
            {
               while (rawTokens.NotEmpty() && int.TryParse(rawTokens.Front().Value, out var _))
               {
                  var tok = searchTokens.InsertBack(rawTokens.RemoveFront()).Back();
                  combinedString += tok.Value;
                  debugSpan += tok.DebugInfo;
               }

               if (rawTokens.NotEmpty() && rawTokens.Front().Value == ".")
               {
                  {
                     var tok = searchTokens.InsertBack(rawTokens.RemoveFront()).Back();
                     combinedString += tok.Value;
                     debugSpan += tok.DebugInfo;
                  }

                  while (rawTokens.NotEmpty() && int.TryParse(rawTokens.Front().Value, out var _))
                  {
                     var tok = searchTokens.InsertBack(rawTokens.RemoveFront()).Back();
                     combinedString += tok.Value;
                     debugSpan += tok.DebugInfo;
                  }
               }

               result.InsertBack(new Token(TokenType.Numeric, combinedString, debugSpan));
               continue;
            }

            while (rawTokens.NotEmpty() && !IsSeparatorToken(rawTokens.Front()))
            {
               var tok = searchTokens.InsertBack(rawTokens.RemoveFront()).Back();
               combinedString += tok.Value;
               debugSpan += tok.DebugInfo;
            }

            result.InsertBack(new Token(combinedString, debugSpan));
            searchTokens.RemoveFront();
         }
      }

      return result;
   }

   public static TokenList Run(string fileName) => Run(fileName, File.ReadAllText(fileName));
}